lib/Dialect/RTL/Ops.cpp

```diff
@@ -657,12 +657,53 @@ void AddOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
         return success();
       }
 
+      // add(..., x, x) -> add(..., mul(x, 2))
+      if (inputs[size - 1] == inputs[size - 2]) {
+        SmallVector<Value, 2> mulInputs;
+
+        auto two = rewriter.create<ConstantOp>(
+            op.getLoc(), 2, op.getType().cast<IntegerType>());
+        mulInputs.push_back(inputs.back());
+        mulInputs.push_back(two);
+
+        auto mulOp = rewriter.create<rtl::MulOp>(op.getLoc(), mulInputs);
+
+        SmallVector<Value, 4> newOperands(inputs.drop_back(/*n=*/2));
+        newOperands.push_back(mulOp);
+        rewriter.replaceOpWithNewOp<AddOp>(op, op.getType(), newOperands);
+        return success();
+      }
+
+      auto mulOp = inputs[size - 1].getDefiningOp<rtl::MulOp>();
+      APInt multiplier;
+
+      // add(..., x, mul(x, c)) -> add(..., mul(x, c + 1))
+      if (mulOp && mulOp.inputs().size() == 2 &&
+          mulOp.inputs()[0] == inputs[size - 2] &&
+          matchPattern(mulOp.inputs()[1], m_RConstant(multiplier))) {
+
+        SmallVector<Value, 2> mulInputs(mulOp.inputs().drop_back());
```

Comment:
mulInputs is exactly 2 entries in size when it is completed.  I'd sink this down below and just use a static array, ala:

```
Value mulInputs[2] = {mulOp.inputs()[0], rhs };
```

Reply:
This doesn't work:
```
       Value factors[2] = {shlOp.lhs(), rhs};
        auto mulOp = rewriter.create<rtl::MulOp>(op.getLoc(), factors);
```
outputs:
```
In file included from ../lib/Dialect/RTL/Ops.cpp:11:
../llvm/mlir/include/mlir/IR/PatternMatch.h:258:11: error: no matching member function for call to 'build'
    OpTy::build(*this, state, args...);
    ~~~~~~^~~~~
...
```

 However, using a std::array does work.